### ISSUE TITLE
Add Vector Data Type

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#2651]
Starting point:

```scala
///
/// Represents a mutable or immutable array with elements of type `a`.
///
/// If `imm` is true then the array is immutable.
///
pub enum Vector[_imm: Bool, a: Type] {
    case Vector(Array[a])
}

///
/// We use `Imm` as a type alias for true (which means immutable).
///
type alias Imm = true

///
/// We use `Mut` as a type alias for false (which means mutable).
///
type alias Mut = false

namespace Vector {

    ///
    /// Returns a fresh vector of length `l` filled with the element `x`.
    ///
    pub def new(x: a, l: Int32): Vector[imm, a] & Impure = Vector([x; l])

    ///
    /// Returns a frozen copy of the vector `v`.
    ///
    /// Freezing a vector is impure because the vector must be copied.
    ///
    /// Note: The vector cannot already be frozen. That would be non-sensical.
    ///
    pub def freeze(v: Vector[Mut, a]): Vector[Imm, a] & Impure = match v {
        case Vector(arr) => 
            let len = arr.length;
            Vector(arr[0 .. len])
    }

    ///
    /// Returns the element at position `i` in the vector `v`.
    ///
    pub def get(v: Vector[imm, a], i: Int32): a & imm = match v {
        case Vector(arr) => arr[i] as & imm
    }

    ///
    /// Stores the value `x` at position `i` in the vector `v`.
    ///
    /// Note: The vector must be mutable.
    ///
    pub def put(v: Vector[Mut, a], i: Int32, x: a): Unit & Impure = match v {
        case Vector(arr) => arr[i] = x
    }

    ///
    /// Returns `Some(x)` if `x` is the first element of `a`.
    ///
    /// Returns `None` if `a` is empty.
    ///
    pub def head(v: Vector[imm, a]): Option[a] & imm = match v {
        case Vector(arr) => 
            if (arr.length == 0) 
                None 
            else 
                Some(arr[0] as & imm)
    }

    ///
    /// Returns the number of elements in `v` that satisfy the predicate `f`.
    ///
    pub def count(f: a -> Bool & ef, v: Vector[imm, a]): Int32 & (ef and imm) = match v {
        case Vector(arr) => countAcc(f, arr, 0, 0)
    } as & (ef and imm)

    def countAcc(f: a -> Bool & ef, a: Array[a], i: Int32, acc: Int32): Int32 & Impure = 
        if (i == a.length)
            acc
        else if (f(a[i]))
            countAcc(f, a, i + 1, acc + 1)
        else 
            countAcc(f, a, i + 1, acc + 0)

    ///
    /// Returns the result of applying `f` to every element in `v`.
    ///
    /// The result is a new array.
    ///
    pub def map(f: a -> b & ef, v: Vector[imm, a]): Vector[imm, b] & (ef and imm) = match v {
        case Vector(src) => 
            let dst = [default; src.length];
            mapHelper(f, src, dst, 0);
            Vector(dst)
    } as & (ef and imm)

    def mapHelper(f: a -> b & ef, src: Array[a], dst: Array[b], i: Int32): Unit & Impure = 
        if (i == src.length) 
            ()
        else {
            dst[i] = f(src[i]);
            mapHelper(f, src, dst, i + 1)
        }

}

def main(_args: Array[String]): Int32 & Impure = 
    let v: Vector[Imm, Int32] = Vector([0, 1, 2, 3] as & Pure);
    let r = Vector.count(x -> x > 0, v);
    println(r);
    0
```

Task list to come:

- TBD

**Comment by magnus-madsen** [comment]
@jaschdoc Do you think you would be interested in this?

And if so, do you think you could describe some milestones ("phases")? Going from easy to harder-- inspired by what is currently in `Array`.

**Comment by jaschdoc** [comment]
Yes, certainly! What's the priority of this? Should we finish Iterator first (or a first draft at least) before moving on to this?

**Comment by magnus-madsen** [comment]
> Yes, certainly! What's the priority of this? Should we finish `Iterator` first (or a first draft at least) before moving on to this?

Low priority. Deadline stuff is passed. I would say error messages, iterator, and then this.

**Comment by jaschdoc** [comment]
# Vector functions

## Phase 1
  - [ ] `isEmpty`
  - [ ] `last`
  - [ ] `range`
  - [ ] `repeat`
  - [ ] `toArray`
  - [ ] `toList`
  - [ ] `toSet`
  - [ ] `toMap`

## Phase 2
  - [ ] `filter`
  - [ ] `mapWithIndex`
  - [ ] `foreach`
  - [ ] `foreachWithIndex`

## Phase 3
  - [ ] `foldLeft`
  - [ ] `foldRight`
  - [ ] `reduceLeft`
  - [ ] `reduceRight`

## Phase 4
  - [ ] `append`
  - [ ] `reverse`
  - [ ] `replace`

## Phase 5
  - [ ] `dropLeft`
  - [ ] `dropRight`
  - [ ] `takeLeft`
  - [ ] `takeRight`
  - [ ] `zip`
  - [ ] `zipWith`

## Phase 6
  - [ ] `indexOf`
  - [ ] `indexOfRight`
  - [ ] `indices`
  - [ ] `findLeft`
  - [ ] `findRight`

## Phase 7
TBD

**Comment by magnus-madsen** [comment]
This is for an experiment. I would like to have Phase 1, 2, and 3 implemented and unit tests (e.g. a 3-5 tests per function. Maybe more to tests both the immutable and mutable cases). We should not proceed to Phase 4 before some other stuff in the compiler has been settled.

**Comment by magnus-madsen** [comment]
@HansenJonathan I think it would be a good exercise if you worked on Phase 1 of this mini project.

**Comment by thinking-tower** [comment]
@magnus-madsen I'm interested at trying this out!

**Comment by magnus-madsen** [comment]
@thinking-tower Cool. Give it a spin! I don't think anyone is working on it anymore :)
